### PR TITLE
Fix exponential query sharding time of binops

### PR DESCRIPTION
**Pull request by colega**
#### What this PR does

When sharding operations like `expr1 or expr2 or ... or expr_n`, the last expression has to be cloned `2^n` times by the `canShardAllVectorSelectors` check.

We can avoid that by caching the results of that function for each expr, making the execution time linear again.

#### Which issue(s) this PR fixes or relates to

Fixes an internal issue.

#### Checklist

- [x] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Review by pracucci — APPROVED** [review]
Thanks! LGTM

File: pkg/frontend/querymiddleware/astmapper/sharding.go [hunk 1]
```diff
@@ -132,17 +136,25 @@ func (summer *shardSummer) MapExpr(expr parser.Expr) (mapped parser.Expr, finish
...
-			c.shards = 1
-			m := NewASTExprMapper(c)
+		canShardAllVectorSelectors := func(expr parser.Expr) (can bool) {
+			query := expr.String()
+			// We need to cache the results of this function to avoid processing it again and again
+			// in queries like `a or b or c or d`, which would lead to exponential processing time.
+			if can, ok := summer.canShardAllVectorSelectorsCache[query]; ok {
+				return can
+			}
+			defer func() { summer.canShardAllVectorSelectorsCache[query] = can }()
```
Comment (krajorama) [inline pkg/frontend/querymiddleware/astmapper/sharding.go:146]: how is this caching the result of the function which is `c.stats.GetShardedQueries() == countVectorSelectors(mappedExpr)`

Comment (colega) [reply pkg/frontend/querymiddleware/astmapper/sharding.go:146]: `can` is a named return value. So `return` statement assigns the value to it, and then the `defer` statement is executed.

Comment (krajorama) [reply pkg/frontend/querymiddleware/astmapper/sharding.go:146]: ugh, magic, thanks

**Pull request merged**